File: lincheck/src/test/java/org/jetbrains/kotlinx/lincheck/test/ExceptionAsResultTest.java
package org.jetbrains.kotlinx.lincheck.test;

/*
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2015 - 2018 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import org.jetbrains.kotlinx.lincheck.LinChecker;
import org.jetbrains.kotlinx.lincheck.annotations.Operation;
import org.jetbrains.kotlinx.lincheck.strategy.stress.StressCTest;
import org.junit.Test;

import java.io.IOException;

@StressCTest(iterations = 1)
public class ExceptionAsResultTest {
    @Operation(handleExceptionsAsResult = NullPointerException.class)
    public void npeIsOk() {
        ((String) null).charAt(0);
    }

    @Operation(handleExceptionsAsResult = Exception.class)
    public void subclassExceptionIsOk() throws Exception {
        throw new IOException();
    }

    @Test
    public void test() {
        LinChecker.check(ExceptionAsResultTest.class);
    }
}

File: src/native/test/ExceptionAsResultTest.kt
import org.jetbrains.kotlinx.lincheck.*
import kotlin.test.*

/*
* #%L
* Lincheck
* %%
* Copyright (C) 2015 - 2018 Devexperts, LLC
* %%
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Lesser Public License for more details.
*
* You should have received a copy of the GNU General Lesser Public
* License along with this program.  If not, see
* <http://www.gnu.org/licenses/lgpl-3.0.html>.
* #L%
*/

class ExceptionAsResultTest {
    fun npeIsOk() {
        (null as String?)!![0]
    }

    fun subclassExceptionIsOk() {
        if ((0..1).random() == 1) throw IndexOutOfBoundsException(MESSAGE) else throw IllegalStateException(MESSAGE)
    }

    @Test
    fun test() {
        try {
            LincheckStressConfiguration<ExceptionAsResultTest>().apply {
                iterations(1)
                requireStateEquivalenceImplCheck(false)

                initialState { ExceptionAsResultTest() }

                operation(ExceptionAsResultTest::npeIsOk, "npeIsOk", listOf(NullPointerException::class))
                operation(ExceptionAsResultTest::subclassExceptionIsOk, "subclassExceptionIsOk", listOf(Throwable::class))
            }.runTest()
            fail("Should fail with AssertionError")
        } catch (e: AssertionError) {
            val m = e.message
            assertTrue(m!!.contains("IllegalStateException") || m.contains("IndexOutOfBoundsException"))
            assertFalse(m.contains(MESSAGE))
        }
    }

    companion object {
        private const val MESSAGE = "iujdhfgilurtybfu"
    }
}